lib/miq_expression.rb

```diff
@@ -323,31 +340,121 @@ def preprocess_exp!(exp)
+  #   sql1, ruby1  | sql1       | ruby1        |
+  #   ruby1, ruby2 |            | ruby1, ruby1 |
+  #
+  def reduce_exp_children(children, mode, swap:)
+    seen = MODE_NONE
+    filtered_children = []
+    children.each do |child|
+      child_exp, child_seen = reduce_exp(child, mode, :swap => swap)
+      seen |= child_seen
+      filtered_children << child_exp if child_exp
+    end
+    [filtered_children, seen]
+  end
+  private :reduce_exp_children
+
+  # Cut up an expression into 2 expressions that can be applied sequentially:
+  # orig_exp == (exp mode sql) AND (exp mode=ruby)
+  #
+  # the sql expression is applied in the db
+  #
+  # @param exp  [Hash]               ast for miq_expression
+  # @param mode [MODE_RUBY|MODE_SQL] whether we are reducing for a sql or ruby generation
+  # @param swap [boolean]            true if we are in a NOT clause and applying Demorgan's law
+  #
+  # @returns [Hash, mode]
+  #   Hash: expression that works for the given mode
+  #   [MODE_SQL|MODE_RUBY|MODE_BOTH]: mode summary for the children
+  #
+  # NOTE on Compound nodes:
+  #
+  # exp             |==>| output (mode=sql) |and| output (mode=ruby)
+  # ----------------|---|-------------------|---|-----------------
+  # sql1  AND sql2  |==>| sql1 AND sql2     |AND|
+  # sql1  AND ruby1 |==>| sql1              |AND| ruby1
+  # ruby1 AND ruby2 |==>|                   |AND| ruby1 AND ruby2
+  #
+  # The AND case uses all nodes that match the input mode
+  #
+  # exp            |==>| output (mode=sql) |and| output (mode=ruby)
+  # ---------------|---|-------------------|---|-----------------
+  # sql1  OR sql2  |==>| sql1 OR sql2      |AND|
+  # sql1  OR ruby1 |==>|                   |AND| sql1 OR ruby1
+  # ruby1 OR ruby2 |==>|                   |AND| ruby1 OR ruby2
+  #
+  # The OR case uses all nodes that match the input mode with one exception:
+  # mixed mode expressions are completely applied in ruby to keep the same logical result.
+  #
+  #
+  # exp                |==>| exp (mode=sql) |and| exp (mode=ruby)
+  # -------------------|---|----------------|---|-----------------
+  # !(sql  OR sql)     |==>| !(sql OR sql)  |AND|
+  # !(sql  OR ruby)    |==>| !(sql)         |AND| !(ruby)
+  # !(ruby1 OR ruby2)  |==>|                |AND| !(ruby1 OR ruby2)
+  #
+  # exp                |==>| exp (mode=sql) |and| exp (mode=ruby)
+  # -------------------|---|----------------|---|-----------------
+  # !(sql   AND sql)   |==>| !(sql AND sql) |AND|
+  # !(sql   AND ruby)  |==>|                |AND| !(sql AND ruby)
+  # !(ruby1 AND ruby2) |==>|                |AND| !(ruby1 AND ruby2)
+  #
+  # Inside a NOT, the OR acts like the AND, and the AND acts like the OR
+  # so follow the AND logic if we are not swapping (and vice versa)
+  #
+  def reduce_exp(exp, mode, swap: false)
     operator = exp.keys.first
-    case operator.downcase
-    when "and"
-      exp[operator].dup.each { |atom| preprocess_for_sql(atom, attrs) }
```

Comment:
Why was this dup here?  Do we need it below?

Reply:
The old `preprocess_for_sql` modified the expression passed in as it worked. Sometimes that was wanted and other times it was not. So we needed to `exp.dup` in some of the operators and not in others. (this got too confusing for me)

Now, we are modifying the tree up front in `preprocess_exp!`. Mostly this is removing the token and reformatting `"5.megabytes"`. We get all the modifications out of the way up front. So the `reduce_exp` phase only removes nodes. And when it does that, it returns a different array (rather than modifying the array in place (i.e.: `exp.delete(operator)`).

I discovered this because the tests started failing in strange ways. You can see all the `freeze` statements I put in there to catch anything strange happening.

In the end, separating this into 2 phases allows us to fix the Demorgan bug and handle the differences between reducing ruby and sql.

Reply:
I think you're saying we don't need the `dup` 🤣